flocker/node/agents/ebs.py

```diff
@@ -243,59 +335,88 @@ def _blockdevicevolume_from_ebs_volume(ebs_volume):
     )
 
 
-def _wait_for_volume(volume,
-                     start_status,
-                     transient_status,
-                     end_status):
+def _get_ebs_volume_state(volume):
+    """
+    Fetch input EBS volume's latest state from backend.
+
+    :param boto.ec2.volume volume: Volume that needs state update.
+
+    :returns: EBS volume with latest state known to backend.
+    :rtype: boto.ec2.volume
+
+    """
+    return volume.update()
+
+
+def _wait_for_volume_state_change(operation,
+                                  volume,
+                                  update=_get_ebs_volume_state,
+                                  timeout=VOLUME_STATE_CHANGE_TIMEOUT):
     """
     Helper function to wait for a given volume to change state
     from ``start_status`` via ``transient_status`` to ``end_status``.
 
-    :param boto.ec2.volume volume: Volume to check
-        status for.
-    :param unicode start_status: Volume status at starting point.
-    :param unicode transient_status: Allowed transient state for
-        volume to be in, on the way to ``end_status``.
-    :param unicode end_status: Expected destination status for
-        the input volume.
+    :param NamedConstant operation: Operation triggering volume state change.
+        A value from ``VolumeOperations``.
+    :param boto.ec2.volume volume: Volume to check status for.
+    :param update: Method to use to fetch EBS volume's latest state.
+    :param int timeout: Seconds to wait for volume operation to succeed.
 
-    :raises Exception: When input volume failed to reach
-        expected destination status.
+    :raises Exception: When input volume fails to reach expected backend
+        state for given operation within timeout seconds.
     """
     # It typically takes a few seconds for anything to happen, so start
     # out sleeping a little before doing initial check to reduce
     # unnecessary polling of the API:
     time.sleep(5.0)
 
-    # Wait ``VOLUME_STATE_CHANGE_TIMEOUT`` seconds for
+    # XXX Catch KeyError.
```

Comment:
Well... you'd get a KeyError if anyone ever passed in an unknown operation. But that's a coding bug that you can't actually handle. So best you can do is log a traceback, which raising KeyError will do anyway. So no need for this comment, or further action, I think.